Add Origin API infrastructure

annevk opened the pull request:
#### 7339d3360824ffc88be7fa8bd74280d5a4b10a75
<pre>
Add Origin API infrastructure
<a href="https://bugs.webkit.org/show_bug.cgi?id=303617">https://bugs.webkit.org/show_bug.cgi?id=303617</a>
<a href="https://rdar.apple.com/165900187">rdar://165900187</a>

Reviewed by Youenn Fablet.

Implement
<a href="https://html.spec.whatwg.org/multipage/browsers.html#the-origin-interface">https://html.spec.whatwg.org/multipage/browsers.html#the-origin-interface</a>
behind the OriginAPIEnabled flag.

Giving MessageEvent and ExtendableMesssageEvent a SecurityOrigin will
be done in a separate change for easy of blame and review.

Canonical link: <a href="https://commits.webkit.org/304163@main">https://commits.webkit.org/304163@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8e34d87733989a4ba5f84095e195d5885a25999

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134757 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7201 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46005 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142277 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86679 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/40d83e71-8109-4812-b79b-08314445c14c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136627 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7053 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/eb8feb22-e5ce-47af-859e-b0b0c5ac8387) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137704 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5503 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120784 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83806 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/23c1f1fc-2c61-4322-bc03-68d756fc1c62) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5347 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2956 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2866 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126795 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38931 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144972 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133264 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6877 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39511 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6947 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5770 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111690 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5182 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117063 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60797 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20799 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6925 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35247 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166169 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6725 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70506 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43431 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6961 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6834 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/html/Origin.h:
```diff
@@ -0,0 +1,53 @@
+
+private:
+    static Ref<Origin> create(Ref<SecurityOrigin>&&);
+    Origin(Ref<SecurityOrigin>&&);
```
```suggestion
    explicit Origin(Ref<SecurityOrigin>&&);
```

youennf commented on Source/WebCore/html/Origin.cpp:
```diff
@@ -0,0 +1,129 @@
+
+namespace WebCore {
+
+inline Origin::Origin(Ref<SecurityOrigin>&& securityOrigin)
```
We usually reserve inline for header.

youennf commented on Source/WebCore/html/Origin.h:
```diff
@@ -0,0 +1,53 @@
+
+class SecurityOrigin;
+
+class Origin final : public RefCountedAndCanMakeWeakPtr<Origin> {
```
I would tend to add CanMakeWeakPtr only when needed, is it already needed?.

youennf commented on Source/WebCore/html/Origin.cpp:
```diff
@@ -0,0 +1,129 @@
+#include "DOMWindow.h"
+#include "ExtendableMessageEvent.h"
+#include "HTMLAnchorElement.h"
+#include "HTMLAreaElement.h"
```
It seems we might not need all those headers.

youennf commented on Source/WebCore/html/Origin.cpp:
```diff
@@ -0,0 +1,129 @@
+#include "JSOrigin.h"
+#include "JSWorkerGlobalScope.h"
+#include "LocalDOMWindow.h"
+#include "MessageEvent.h"
```
If we include JSMessageEvent.h, we do no need MessageEvent.h

youennf commented on Source/WebCore/html/Origin.cpp:
```diff
@@ -0,0 +1,129 @@
+    return adoptRef(*new Origin(WTFMove(securityOrigin)));
+}
+
+ExceptionOr<Ref<Origin>> Origin::from(ScriptExecutionContext& context, JSC::JSValue value)
```
Could you add a link to the spec?

youennf commented on Source/WebCore/html/Origin.cpp:
```diff
@@ -0,0 +1,129 @@
+    if (value.isString()) {
+        auto parsedURL = URL { value.getString(context.globalObject()) };
+        if (parsedURL.isValid())
+            return create(SecurityOrigin::create(parsedURL));
```
I would tend to write it as:
```
if (value.isString()) {
    if (!parsedURL.isValid())
        return Exception { ExceptionCode::TypeError, "URL is invalid"_s };
   return create(...);
}
if (auto...)
```
This is closer to the coding style of returning early and adding descriptions to exceptions is a good practice.

youennf commented on Source/WebCore/html/Origin.cpp:
```diff
@@ -0,0 +1,129 @@
+    return m_origin->isOpaque();
+}
+
+bool Origin::isSameOrigin(Origin& other) const
```
```suggestion
bool Origin::isSameOrigin(const Origin& other) const
```

youennf commented on Source/WebCore/html/Origin.cpp:
```diff
@@ -0,0 +1,129 @@
+    return m_origin->isSameOriginAs(other.m_origin);
+}
+
+bool Origin::isSameSite(Origin& other) const
```
```suggestion
bool Origin::isSameSite(const Origin& other) const
```

annevk committed:
Add Origin API infrastructure
https://bugs.webkit.org/show_bug.cgi?id=303617
rdar://165900187

Reviewed by Youenn Fablet.

Implement
https://html.spec.whatwg.org/multipage/browsers.html#the-origin-interface
behind the OriginAPIEnabled flag.

Giving MessageEvent and ExtendableMesssageEvent a SecurityOrigin will
be done in a separate change for easy of blame and review.

Canonical link: https://commits.webkit.org/304163@main
